CSS transform functions should not eagerly evaluate calc()

weinig opened the pull request:
#### 8531e44d52ced52c67b8488e359d797577f86b2f
<pre>
CSS transform functions should not eagerly evaluate calc()
<a href="https://bugs.webkit.org/show_bug.cgi?id=278993">https://bugs.webkit.org/show_bug.cgi?id=278993</a>

Reviewed by Darin Adler.

Refactors CSS transform parsing and style building code. Moves
parsing into new CSSPropertyParserConsumer+Transform.h/cpp files
and style building into new TransformOperationsBuilder.h/cpp,
matching the pattern used for filters.

To make things a bit more structured, each individual transform
function has been given their own parsing and style building
functions, consumeFooFunctionArguments and createFooTransformOperation
respectively, each with a spec link and grammar definition.
This makes it really clear and will help with the incremental
move to toward generated versions in the future.

A few uses of &quot;deprecated&quot; primitive value resolvers were removed
from `consumeScale` and `consumeRotate`, replaced with calc()
aware accessors.

Following the precedent set by color parsing, and more recently
filter parsing, a &quot;raw&quot; parsing from string function was added
for transforms to allow `DOMMatrixReadOnly` to parse a transform
without spinning up all the CSSParser / MutableStyleProperties
infrastructure. To keep the behavior that this fails if the
parsed value contains relative lengths, the code to check if
dependencies can be resolved by a specific instance of
`CSSToLengthConversionData` was factored out of CSSPrimitiveValue
and into CSSValue and used in parseTransformRaw.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-animation-transform-none.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-transition-transform-list-none.tentative-expected.txt:
    - Update results for now passing tests.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
* Source/WebCore/css/TransformFunctions.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/style/FilterOperationsBuilder.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/TransformOperationsBuilder.cpp: Renamed from Source/WebCore/css/TransformFunctions.cpp.
* Source/WebCore/style/TransformOperationsBuilder.h: Renamed from Source/WebCore/css/TransformFunctions.h.

Canonical link: <a href="https://commits.webkit.org/283074@main">https://commits.webkit.org/283074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb202ba67e731b1581a67bbd4c5609cb9d0b63ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65140 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44507 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69164 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67258 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52290 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16028 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52332 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10890 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68206 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41148 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56387 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32956 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37819 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13758 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14622 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14097 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70869 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9092 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56447 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14363 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7514 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1189 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40319 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42577 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41140 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/DOMMatrixReadOnly.cpp:
```diff
@@ -227,21 +228,15 @@ ExceptionOr<DOMMatrixReadOnly::AbstractMatrix> DOMMatrixReadOnly::parseStringInt
 
-    auto styleDeclaration = MutableStyleProperties::create();
-    if (CSSParser::parseValue(styleDeclaration, CSSPropertyTransform, string, IsImportant::Yes, HTMLStandardMode) == CSSParser::ParseResult::Error)
+    CSSToLengthConversionData conversionData { };
```
Surprised the braces are needed here. Should we initialize more in the `struct` or `class` itself instead?

weinig replied:
I don't think they are needed. I'll try to remove them.

weinig committed:
CSS transform functions should not eagerly evaluate calc()
https://bugs.webkit.org/show_bug.cgi?id=278993

Reviewed by Darin Adler.

Refactors CSS transform parsing and style building code. Moves
parsing into new CSSPropertyParserConsumer+Transform.h/cpp files
and style building into new TransformOperationsBuilder.h/cpp,
matching the pattern used for filters.

To make things a bit more structured, each individual transform
function has been given their own parsing and style building
functions, consumeFooFunctionArguments and createFooTransformOperation
respectively, each with a spec link and grammar definition.
This makes it really clear and will help with the incremental
move to toward generated versions in the future.

A few uses of "deprecated" primitive value resolvers were removed
from `consumeScale` and `consumeRotate`, replaced with calc()
aware accessors.

Following the precedent set by color parsing, and more recently
filter parsing, a "raw" parsing from string function was added
for transforms to allow `DOMMatrixReadOnly` to parse a transform
without spinning up all the CSSParser / MutableStyleProperties
infrastructure. To keep the behavior that this fails if the
parsed value contains relative lengths, the code to check if
dependencies can be resolved by a specific instance of
`CSSToLengthConversionData` was factored out of CSSPrimitiveValue
and into CSSValue and used in parseTransformRaw.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-animation-transform-none.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/animation/custom-property-transition-transform-list-none.tentative-expected.txt:
    - Update results for now passing tests.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/DOMMatrixReadOnly.cpp:
* Source/WebCore/css/TransformFunctions.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/style/FilterOperationsBuilder.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/TransformOperationsBuilder.cpp: Renamed from Source/WebCore/css/TransformFunctions.cpp.
* Source/WebCore/style/TransformOperationsBuilder.h: Renamed from Source/WebCore/css/TransformFunctions.h.

Canonical link: https://commits.webkit.org/283074@main
